crates/red_knot_server/src/server/api/requests/inlay_hints.rs

```diff
@@ -0,0 +1,78 @@
+use std::borrow::Cow;
+
+use crate::server::api::traits::{BackgroundDocumentRequestHandler, RequestHandler};
+use crate::server::client::Notifier;
+use crate::DocumentSnapshot;
+use lsp_types::request::InlayHintRequest;
+use lsp_types::{InlayHintParams, Url};
+use red_knot_ide::get_inlay_hints;
+use red_knot_project::Db;
+use red_knot_project::ProjectDatabase;
+use ruff_db::source::{line_index, source_text};
+use ruff_text_size::Ranged;
+
+pub(crate) struct InlayHintRequestHandler;
+
+impl RequestHandler for InlayHintRequestHandler {
+    type RequestType = InlayHintRequest;
+}
+
+impl BackgroundDocumentRequestHandler for InlayHintRequestHandler {
+    fn document_url(params: &InlayHintParams) -> Cow<Url> {
+        Cow::Borrowed(&params.text_document.uri)
+    }
+
+    fn run_with_snapshot(
+        snapshot: DocumentSnapshot,
+        db: ProjectDatabase,
+        _notifier: Notifier,
+        params: InlayHintParams,
+    ) -> crate::server::Result<Option<Vec<lsp_types::InlayHint>>> {
+        let Some(file) = snapshot.file(&db) else {
+            tracing::debug!("Failed to resolve file for {:?}", params);
+            return Ok(None);
+        };
+
+        let editor_options = db
+            .project()
+            .metadata(&db)
+            .options()
+            .editor
+            .clone()
+            .unwrap_or_default();
+
+        if !editor_options.inlay_hints.unwrap_or(false) {
+            return Ok(None);
+        }
+
+        let inlay_hints = get_inlay_hints(&db, file);
```

Comment:
The `InlayHintRequestHandler` provide a `range` to avoid computing the inlay hints for the entire document. We should respect that parameter